test/test_charm.py

```diff
@@ -419,45 +389,94 @@ def test_relations_meta(self):
+    assert meta.requires['metrics'].interface_name == 'prometheus-scraping'
+    assert meta.requires['metrics'].limit is None
+    assert meta.requires['metrics'].scope == 'global'  # Default value
+    assert meta.requires['metrics'].optional
 
-    def test_relations_meta_limit_type_validation(self):
-        with pytest.raises(TypeError, match=r"limit should be an int, not <class 'str'>"):
-            # language=YAML
-            self.meta = ops.CharmMeta.from_yaml('''
+
+def test_relations_meta_limit_type_validation():
+    with pytest.raises(TypeError, match=r"limit should be an int, not <class 'str'>"):
+        # language=YAML
+        ops.CharmMeta.from_yaml('''
 name: my-charm
 requires:
   database:
     interface: mongodb
     limit: foobar
 ''')
 
-    def test_relations_meta_scope_type_validation(self):
-        with pytest.raises(
-            TypeError,
-            match="scope should be one of 'global', 'container'; not 'foobar'"
-        ):
-            # language=YAML
-            self.meta = ops.CharmMeta.from_yaml('''
+
+def test_relations_meta_scope_type_validation():
+    with pytest.raises(
+        TypeError,
+        match="scope should be one of 'global', 'container'; not 'foobar'"
+    ):
+        # language=YAML
+        ops.CharmMeta.from_yaml('''
 name: my-charm
 requires:
   database:
     interface: mongodb
     scope: foobar
 ''')
 
-    @classmethod
-    def _get_action_test_meta(cls):
-        # language=YAML
-        return ops.CharmMeta.from_yaml(metadata='''
+
+def test_meta_from_charm_root():
+    with tempfile.TemporaryDirectory() as d:
+        td = pathlib.Path(d)
+        (td / 'metadata.yaml').write_text(
+            yaml.safe_dump(
+                {"name": "bob",
+                    "requires": {
+                        "foo":
+                            {"interface": "bar"}
+                    }}))
+        meta = ops.CharmMeta.from_charm_root(td)
+        assert meta.name == "bob"
+        assert meta.requires['foo'].interface_name == "bar"
+
+
+def test_actions_from_charm_root():
+    with tempfile.TemporaryDirectory() as d:
+        td = pathlib.Path(d)
+        (td / 'actions.yaml').write_text(
+            yaml.safe_dump(
+                {"foo": {
+                    "description": "foos the bar",
+                    "additionalProperties": False
+                }}
+            )
+        )
+        (td / 'metadata.yaml').write_text(
+            yaml.safe_dump(
+                {"name": "bob",
+                    "requires": {
+                        "foo":
+                            {"interface": "bar"}
+                    }}))
+
+        meta = ops.CharmMeta.from_charm_root(td)
+        assert meta.name == "bob"
+        assert meta.requires['foo'].interface_name == "bar"
+        assert not meta.actions['foo'].additional_properties
+        assert meta.actions['foo'].description == "foos the bar"
+
+
+def _setup_test_action(fake_script: typing.Callable[..., None]):
```

Comment:
I think it would be clearer (and a simpler type) if `fake_script` was a `FakeScript` here, and we used `fake_script.write()` in the function, like the other calls.